mdp/nodes/recursive_expansion_nodes.py

```diff
@@ -0,0 +1,623 @@
+            the function sequence selected is defined or orthogonal.
+
+        :param x: The data to be expanded. Observations/samples must
+            be along the first axis, different variables along the second.
+        :type x: numpy.ndarray
+
+        :param prec: (Numerical) tolerance when checking validity.
+        :type prec: float
+
+        :raise mdp.NodeException: If one or more values lie outside of the function
+            specific domain.
+        """
+        xmax = np.amax(x)-prec
+        xmin = np.amin(x)+prec
+
+        if (self.upper < xmax) or (self.lower > xmin):
+            raise mdp.NodeException(
+                "One or more values lie outside of the function specific domain.")
+
+    @staticmethod
+    def is_trainable():
+        return False
+
+    @staticmethod
+    def is_invertible():
+        return False
+
+
+class TrainableRecursiveExpansionNode(RecursiveExpansionNode):
+    """Recursively computable (orthogonal) expansions and a
+    trainable transformation to the domain of the expansions.
+
+    .. attribute:: lower
+
+        The lower bound of the domain, on which the recursion
+        function sequence selected, is defined or orthogonal.
+
+    .. attribute:: upper
+
+        The upper bound of the domain, on which the recursion
+        function sequence selected, is defined or orthogonal.
+    """
+
+    def __init__(self, degree=1, recf='standard_poly', check=True, with0=True,
+                 input_dim=None, dtype=None):
+        super(TrainableRecursiveExpansionNode, self).__init__(degree,
+                                                              recf=recf, check=check,
+                                                              with0=with0, input_dim=input_dim,
+                                                              dtype=dtype)
+
+        self.amaxcolumn = None
+        self.amincolumn = None
+        self.amax = None
+        self.amin = None
+
+    def _execute(self, x):
+        """Apply the transformation and execute RecursiveExpansionNode.
+
+        :param x: The data to be expanded. Observations/samples must
+            be along the first axis, different variables along the second.
+        :type x: numpy.ndarray
+        """
+        self.domain_transformation(x)
+        return super(TrainableRecursiveExpansionNode, self)._execute(x)
+
+    def _train(self, x):
+        """Determine coordinatewise and absolute maxima and minima.
+
+        :param x: Chuck of data to be used for training. Observations/samples
+            must be along the first axis, different variables along the second.
+        :type x: numpy.ndarray
+
+        The values are used to generate a transformation to the valid domain
+        for the data.
+        """
+        if self.amin is None:
+            self.amaxcolumn = np.amax(x, axis=0)
+            self.amincolumn = np.amin(x, axis=0)
+            self.amax = np.amax(x)
```

Comment:
Can't this be computed as np.max(self.amaxcolumn)? No need to search the whole chunk twice. Searching amaxcolumn is much cheaper. Same for min.

Reply:
Fixed. Thanks for the review!